<h1 align="center"><b>CS116.M11 - LẬP TRÌNH PYTHON CHO MÁY HỌC</b></h1>
<h1 align="center"><b>BÀI TẬP 7 - NGÀY 19/10/2021</b></h1>

# **Thông tin sinh viên**

1. ***MSSV:*** 19520214
2. ***Họ tên:*** Cao Hưng Phú
3. ***Bài tập:*** Bài tập 7
4. ***Ngày thực hiện:*** 19/10/2021

# **Bài làm**

### **Các hàm và thư viện cần thiết**

In [28]:
import numpy as np
import pandas as pd
import warnings
warnings.simplefilter("ignore")

### **Bước 1: Load dữ liệu Social Network Ads**

In [22]:
# Tải file Social_Network_Ads.csv từ GoogleDrive
!gdown --id "1vlfarV6qq7ot4KXb9NW8RoG2Rm1cJ0E2"

# Gán path file Social_Network_Ads.csv
path_data = "/content/Social_Network_Ads.csv"

# Đọc file Wine.csv
df = pd.read_csv(path_data)

# Hiển thị 5 dòng đầu của data
df.head(5)

Downloading...
From: https://drive.google.com/uc?id=1vlfarV6qq7ot4KXb9NW8RoG2Rm1cJ0E2
To: /content/Social_Network_Ads.csv
100% 4.90k/4.90k [00:00<00:00, 8.80MB/s]


Age  EstimatedSalary  Purchased
0   19            19000          0
1   35            20000          0
2   26            43000          0
3   27            57000          0
4   19            76000          0

### **Bước 2: Chuẩn hóa dữ liệu với Standard Scaler**

In [42]:
X = np.array(df[['Age', 'EstimatedSalary']])
y = np.array(df['Purchased'])

In [43]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)

### **Bước 3: Chia ra làm 2 tập train và test với K-Fold cross validation, K = 5**

In [56]:
from sklearn.model_selection import KFold

kFold = KFold(n_splits=5)
kFold.get_n_splits(X)

list_X_train, list_y_train, list_X_test, list_y_test = [], [], [], []

for i_train, i_test in kFold.split(X):
  list_X_train.append(X[i_train])
  list_y_train.append(y[i_train])
  list_X_test.append(X[i_test])
  list_y_test.append(y[i_test])

### **Bước 4: Áp dụng lần lượt một trong các classifier sau: KNN, SVM, Decision Tree, Bayesian, Logistic Regression**

#### **Hàm chính để gọi lấy kết quả**

In [64]:
def getResultModel(model):
  list_accuracy = []
  for i in range(len(list_X_train)):
    model.fit(list_X_train[i], list_y_train[i])
    list_accuracy.append(model.score(list_X_test[i], list_y_test[i]))
  return list_accuracy

#### **4.1. KNN**

In [74]:
from sklearn.neighbors import KNeighborsClassifier
KNNmodel = KNeighborsClassifier(n_neighbors=3)
KNNres = getResultModel(KNNmodel)

#### **4.2. SVM**

##### **4.2.1. SVM Linear**

In [140]:
from sklearn.svm import SVC
SVMlinearmodel = SVC(kernel='linear')
SVMlinearres = getResultModel(SVMlinearmodel)

##### **4.2.2. SVM Poly**

In [142]:
from sklearn.svm import SVC
SVMpolymodel = SVC(kernel='poly')
SVMpolyres = getResultModel(SVMpolymodel)

##### **4.2.3. SVM rbf**

In [144]:
from sklearn.svm import SVC
SVMrbfmodel = SVC(kernel='rbf')
SVMrbfres = getResultModel(SVMrbfmodel)

#### **4.3. Decision Tree**

In [76]:
from sklearn.tree import DecisionTreeClassifier
DTmodel = DecisionTreeClassifier()
DTres = getResultModel(DTmodel)

#### **4.4. Bayesian**

In [77]:
from sklearn.naive_bayes import GaussianNB
NBmodel = GaussianNB()
NBres = getResultModel(NBmodel)

#### **4.5. Logistic Regression**

In [78]:
from sklearn.linear_model import LogisticRegression
LRmodel = LogisticRegression()
LRres = getResultModel(LRmodel)

### **Bước 5: Tính accuracy sử dụng KFold**

In [88]:
def showResult(nameModel, listRes):
  print("=> Model: {}".format(nameModel))
  print("* Chi tiết các Fold:")
  for i, res in enumerate(listRes):
    print(" > Fold {} - Score: {}".format(i, res))
  print("* Tổng thể các Fold:")
  print(" > Score {:.2f} (Độ lệch: +-{:.2f})".format(np.mean(listRes), np.std(listRes)))

In [89]:
# KNN
showResult('KNN', KNNres)

=> Model: KNN
* Chi tiết các Fold:
 > Fold 0 - Score: 0.95
 > Fold 1 - Score: 0.95
 > Fold 2 - Score: 0.85
 > Fold 3 - Score: 0.8625
 > Fold 4 - Score: 0.85
* Tổng thể các Fold:
 > Score 0.89 (Độ lệch: +-0.05)


In [146]:
# SVM
showResult('SVM linear', SVMlinearres)
print("\n")
showResult('SVM poly', SVMpolyres)
print("\n")
showResult('SVM rbf', SVMrbfres)

=> Model: SVM linear
* Chi tiết các Fold:
 > Fold 0 - Score: 0.7875
 > Fold 1 - Score: 0.9125
 > Fold 2 - Score: 0.825
 > Fold 3 - Score: 0.8375
 > Fold 4 - Score: 0.7375
* Tổng thể các Fold:
 > Score 0.82 (Độ lệch: +-0.06)


=> Model: SVM poly
* Chi tiết các Fold:
 > Fold 0 - Score: 0.8375
 > Fold 1 - Score: 0.95
 > Fold 2 - Score: 0.825
 > Fold 3 - Score: 0.7625
 > Fold 4 - Score: 0.675
* Tổng thể các Fold:
 > Score 0.81 (Độ lệch: +-0.09)


=> Model: SVM rbf
* Chi tiết các Fold:
 > Fold 0 - Score: 0.95
 > Fold 1 - Score: 0.975
 > Fold 2 - Score: 0.825
 > Fold 3 - Score: 0.85
 > Fold 4 - Score: 0.9375
* Tổng thể các Fold:
 > Score 0.91 (Độ lệch: +-0.06)


In [92]:
# Decision Tree
showResult('Decision Tree', DTres)

=> Model: Decision Tree
* Chi tiết các Fold:
 > Fold 0 - Score: 0.8875
 > Fold 1 - Score: 0.95
 > Fold 2 - Score: 0.825
 > Fold 3 - Score: 0.8
 > Fold 4 - Score: 0.8375
* Tổng thể các Fold:
 > Score 0.86 (Độ lệch: +-0.05)


In [93]:
# Bayesian
showResult('Bayesian', NBres)

=> Model: Bayesian
* Chi tiết các Fold:
 > Fold 0 - Score: 0.9
 > Fold 1 - Score: 0.9375
 > Fold 2 - Score: 0.825
 > Fold 3 - Score: 0.825
 > Fold 4 - Score: 0.8
* Tổng thể các Fold:
 > Score 0.86 (Độ lệch: +-0.05)


In [95]:
# Logistic Regression
showResult('Logistic Regression', LRres)

=> Model: Logistic Regression
* Chi tiết các Fold:
 > Fold 0 - Score: 0.7875
 > Fold 1 - Score: 0.9125
 > Fold 2 - Score: 0.825
 > Fold 3 - Score: 0.8375
 > Fold 4 - Score: 0.75
* Tổng thể các Fold:
 > Score 0.82 (Độ lệch: +-0.05)


### **Bước 6: Lập bảng so sánh các phương pháp**

In [136]:
def convertListRes(nameRes, listRes):
  res = [nameRes]
  for x in listRes:
    res.append(x)
  res.append(float("{:.2f}".format(np.mean(listRes))))
  return res

def convertColumnName(dataShowConvert):
  res = ['Tên model']
  for i in range(len(dataShowConvert[0]) - 2):
    res.append("Fold {}".format(i + 1))
  res.append('Trung bình')
  return res

In [147]:
dataShow = [['KNN', KNNres], ['SVM linear', SVMlinearres], ['SVM poly', SVMpolyres], ['SVM rbf', SVMrbfres],\
            ['Decision Tree', DTres], ['Bayesian', NBres], ['Logistic Regression', LRres]]
dataShowConvert = [convertListRes(*x) for x in dataShow]
# dataShowConvert

[['KNN', 0.95, 0.95, 0.85, 0.8625, 0.85, 0.89],
 ['SVM linear', 0.7875, 0.9125, 0.825, 0.8375, 0.7375, 0.82],
 ['SVM poly', 0.8375, 0.95, 0.825, 0.7625, 0.675, 0.81],
 ['SVM rbf', 0.95, 0.975, 0.825, 0.85, 0.9375, 0.91],
 ['Decision Tree', 0.8875, 0.95, 0.825, 0.8, 0.8375, 0.86],
 ['Bayesian', 0.9, 0.9375, 0.825, 0.825, 0.8, 0.86],
 ['Logistic Regression', 0.7875, 0.9125, 0.825, 0.8375, 0.75, 0.82]]

In [151]:
import pandas as pd
from markdownTable import markdownTable
df_show = pd.DataFrame(dataShowConvert, columns=convertColumnName(dataShowConvert))
# df_show.to_markdown()

|    | Tên model           |   Fold 1 |   Fold 2 |   Fold 3 |   Fold 4 |   Fold 5 |   Trung bình |
|---:|:--------------------|---------:|---------:|---------:|---------:|---------:|-------------:|
|  0 | KNN                 |   0.95   |   0.95   |    0.85  |   0.8625 |   0.85   |         0.89 |
|  1 | SVM linear          |   0.7875 |   0.9125 |    0.825 |   0.8375 |   0.7375 |         0.82 |
|  2 | SVM poly            |   0.8375 |   0.95   |    0.825 |   0.7625 |   0.675  |         0.81 |
|  3 | SVM rbf             |   0.95   |   0.975  |    0.825 |   0.85   |   0.9375 |         0.91 |
|  4 | Decision Tree       |   0.8875 |   0.95   |    0.825 |   0.8    |   0.8375 |         0.86 |
|  5 | Bayesian            |   0.9    |   0.9375 |    0.825 |   0.825  |   0.8    |         0.86 |
|  6 | Logistic Regression |   0.7875 |   0.9125 |    0.825 |   0.8375 |   0.75   |         0.82 |